File: android/app/src/main/java/app/offlinerailwaymap/map/StyleBuilder.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Joel Haasnoot

package app.offlinerailwaymap.map

import android.content.Context
import android.util.Log
import app.offlinerailwaymap.data.InstalledPack
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.Calendar

/**
 * Builds the MapLibre style JSON shown on the map.
 *
 * The upstream OpenRailwayMap style is a single style that switches between "Infrastructure",
 * "Speed", ... via `global-state` expressions. MapLibre Native does not evaluate those, so we
 * substitute the chosen state values into the expressions ourselves, evaluate each layer's
 * `visibility` expression to a constant and drop hidden layers.
 *
 * Every installed country pack gets its own copy of the sources and layers, pointing at the
 * PMTiles files on disk.
 */
object StyleBuilder {
    private const val TAG = "StyleBuilder"

    private var ormStyle: JSONObject? = null
    private var basemapStyle: JSONObject? = null

    @Synchronized
    private fun load(context: Context): Pair<JSONObject, JSONObject> {
        if (ormStyle == null) {
            ormStyle = JSONObject(context.assets.open("style/orm-style.json").bufferedReader().readText())
            basemapStyle = JSONObject(context.assets.open("style/basemap-style.json").bufferedReader().readText())
        }
        return ormStyle!! to basemapStyle!!
    }

    fun build(context: Context, mode: MapMode, options: MapOptions, packs: List<InstalledPack>): String {
        val (orm, base) = load(context)
        val state = globalState(orm, mode, options)

        val out = JSONObject()
        out.put("version", 8)
        out.put("name", "OpenRailwayMap offline · ${mode.label}")
        out.put("sprite", orm.getString("sprite"))
        out.put("glyphs", orm.getString("glyphs"))
        val sources = JSONObject()
        val layers = JSONArray()

        // --- Basemap (one source per pack that ships one) ---
        val basemapSource = base.getJSONObject("sources").getJSONObject("basemap")
        val basemapPacks = packs.filter { it.basemapFile != null }
        for (pack in basemapPacks) {
            val src = JSONObject(basemapSource.toString())
            src.put("url", pmtilesUrl(pack.basemapFile!!))
            sources.put("basemap__${pack.info.id}", src)
        }
        val baseLayers = base.getJSONArray("layers")
        for (i in 0 until baseLayers.length()) {
            val layer = baseLayers.getJSONObject(i)
            if (!layer.has("source")) {
                layers.put(JSONObject(layer.toString()))
                continue
            }
            val text = layer.toString()
            for (pack in basemapPacks) {
                val copy = JSONObject(text)
                copy.put("id", "${layer.getString("id")}__${pack.info.id}")
                copy.put("source", "basemap__${pack.info.id}")
                layers.put(copy)
            }
        }

        // --- OpenRailwayMap ---
        val ormSources = orm.getJSONObject("sources")
        val vectorSources = ormSources.keys().asSequence()
            .filter { ormSources.getJSONObject(it).optString("type") == "vector" }
            .toSet()
        for (pack in packs) {
            for (name in vectorSources) {
                val src = JSONObject(ormSources.getJSONObject(name).toString())
                src.put("url", pmtilesUrl(pack.railwayFile))
                sources.put("${name}__${pack.info.id}", src)
            }
        }
        val ormLayers = orm.getJSONArray("layers")
        var hidden = 0
        for (i in 0 until ormLayers.length()) {
            val layer = ormLayers.getJSONObject(i)
            val sourceName = layer.optString("source")
            if (sourceName !in vectorSources) {
                continue
            }
            val substituted = simplify(substitute(layer, state)) as JSONObject
            val layout = substituted.optJSONObject("layout")
            val visibility = layout?.opt("visibility")
            if (visibility is JSONArray) {
                Log.w(TAG, "visibility of ${layer.optString("id")} did not reduce to a constant: $visibility")
                layout.put("visibility", "visible")
            }
            if (layout?.optString("visibility") == "none") {
                hidden++
                continue
            }
            // Filters that reduced to a constant: drop the layer or drop the filter.
            val filter = substituted.opt("filter")
            if (filter is Boolean) {
                if (!filter) {
                    hidden++
                    continue
                }
                substituted.remove("filter")
            }
            val text = substituted.toString()
            for (pack in packs) {
                val copy = JSONObject(text)
                copy.put("id", "${layer.getString("id")}__${pack.info.id}")
                copy.put("source", "${sourceName}__${pack.info.id}")
                layers.put(copy)
            }
        }
        out.put("sources", sources)
        out.put("layers", layers)
        Log.i(TAG, "built style '${mode.id}' for ${packs.size} pack(s): ${layers.length()} layers, $hidden upstream layers hidden")
        return out.toString()
    }

    private fun pmtilesUrl(file: File): String = "pmtiles://file://" + file.absolutePath

    private fun globalState(orm: JSONObject, mode: MapMode, options: MapOptions): Map<String, Any?> {
        val state = HashMap<String, Any?>()
        val defaults = orm.optJSONObject("state")
        if (defaults != null) {
            for (key in defaults.keys()) {
                state[key] = defaults.getJSONObject(key).opt("default")
            }
        }
        state["style"] = mode.id
        state.putAll(mode.globalState)
        state.putAll(options.toGlobalState())
        state["theme"] = "light"
        state["pitched"] = false
        state["bearing"] = 0
        state["hillshade"] = false
        state["openHistoricalMap"] = false
        state["date"] = Calendar.getInstance().get(Calendar.YEAR)
        state["allDates"] = false
        return state
    }

    /** Deep-copies [value], replacing every `["global-state", key]` with the value of that key. */
    private fun substitute(value: Any?, state: Map<String, Any?>): Any? = when (value) {
        is JSONArray -> {
            if (value.length() == 2 && value.opt(0) == "global-state") {
                state[value.getString(1)] ?: JSONObject.NULL
            } else {
                val arr = JSONArray()
                for (i in 0 until value.length()) {
                    arr.put(substitute(value.opt(i), state))
                }
                arr
            }
        }
        is JSONObject -> {
            val obj = JSONObject()
            for (key in value.keys()) {
                obj.put(key, substitute(value.opt(key), state))
            }
            obj
        }
        else -> value
    }

    /**
     * Constant-folds sub-expressions whose operands are all constants (after `global-state`
     * substitution), so that e.g. `["==", "usage", "usage"]` becomes `true` and `["all", ...]`,
     * `["any", ...]` and `["case", ...]` collapse. Besides shrinking the style, this is required
     * because MapLibre Native parses a filter such as `["==", "a", "b"]` as a *legacy* filter
     * (property "a" equals "b") and rejects `["==", true, false]` outright.
     */
    private fun simplify(value: Any?): Any? {
        if (value is JSONObject) {
            val obj = JSONObject()
            for (key in value.keys()) {
                obj.put(key, simplify(value.opt(key)))
            }
            return obj
        }
        if (value !is JSONArray || value.length() == 0 || value.opt(0) !is String) {
            return value
        }
        val op = value.getString(0)
        if (op == "literal") {
            return value
        }
        val args = (1 until value.length()).map { simplify(value.opt(it)) }
        fun rebuild(): JSONArray = JSONArray().put(op).also { arr -> args.forEach { arr.put(it ?: JSONObject.NULL) } }
        when (op) {
            "all" -> {
                val remaining = ArrayList<Any?>()
                for (a in args) {
                    if (a == false) {
                        return false
                    }
                    if (a != true) {
                        remaining.add(a)
                    }
                }
                return when (remaining.size) {
                    0 -> true
                    1 -> remaining[0]
                    else -> JSONArray().put("all").also { arr -> remaining.forEach { arr.put(it) } }
                }
            }
            "any" -> {
                val remaining = ArrayList<Any?>()
                for (a in args) {
                    if (a == true) {
                        return true
                    }
                    if (a != false) {
                        remaining.add(a)
                    }
                }
                return when (remaining.size) {
                    0 -> false
                    1 -> remaining[0]
                    else -> JSONArray().put("any").also { arr -> remaining.forEach { arr.put(it) } }
                }
            }
            "case" -> {
                val kept = ArrayList<Any?>()
                var i = 0
                while (i + 1 < args.size) {
                    val cond = args[i]
                    if (cond == true) {
                        return if (kept.isEmpty()) args[i + 1] else finishCase(kept, args[i + 1])
                    }
                    if (cond != false && !(cond == null || cond === JSONObject.NULL)) {
                        kept.add(cond)
                        kept.add(args[i + 1])
                    }
                    i += 2
                }
                val fallback = args.last()
                return if (kept.isEmpty()) fallback else finishCase(kept, fallback)
            }
            "!" -> return if (args[0] is Boolean) !(args[0] as Boolean) else rebuild()
            "==", "!=", "<", "<=", ">", ">=", "in", "match", "coalesce" -> {
                val allConstant = args.all { it !is JSONArray || (it.length() > 0 && it.opt(0) == "literal") }
                if (allConstant) {
                    return try {
                        evaluate(rebuild())
                    } catch (e: Exception) {
                        rebuild()
                    }
                }
                return rebuild()
            }
            else -> return rebuild()
        }
    }

    private fun finishCase(pairs: List<Any?>, fallback: Any?): JSONArray {
        val arr = JSONArray().put("case")
        pairs.forEach { arr.put(it) }
        arr.put(fallback ?: JSONObject.NULL)
        return arr
    }

    // --- A very small expression evaluator, enough for the upstream `visibility` expressions ---

    private fun evaluate(e: Any?): Any? {
        if (e !is JSONArray || e.length() == 0 || e.opt(0) !is String) {
            return norm(e)
        }
        return when (val op = e.getString(0)) {
            "literal" -> e.opt(1)
            "case" -> {
                var i = 1
                while (i + 1 < e.length()) {
                    if (truthy(evaluate(e.opt(i)))) {
                        return evaluate(e.opt(i + 1))
                    }
                    i += 2
                }
                evaluate(e.opt(e.length() - 1))
            }
            "all" -> (1 until e.length()).all { truthy(evaluate(e.opt(it))) }
            "any" -> (1 until e.length()).any { truthy(evaluate(e.opt(it))) }
            "!" -> !truthy(evaluate(e.opt(1)))
            "==" -> eq(evaluate(e.opt(1)), evaluate(e.opt(2)))
            "!=" -> !eq(evaluate(e.opt(1)), evaluate(e.opt(2)))
            "<" -> compare(evaluate(e.opt(1)), evaluate(e.opt(2))) < 0
            "<=" -> compare(evaluate(e.opt(1)), evaluate(e.opt(2))) <= 0
            ">" -> compare(evaluate(e.opt(1)), evaluate(e.opt(2))) > 0
            ">=" -> compare(evaluate(e.opt(1)), evaluate(e.opt(2))) >= 0
            "in" -> {
                val needle = evaluate(e.opt(1))
                when (val haystack = evaluate(e.opt(2))) {
                    is JSONArray -> (0 until haystack.length()).any { eq(needle, norm(haystack.opt(it))) }
                    is String -> needle is String && haystack.contains(needle)
                    else -> false
                }
            }
            "match" -> {
                val input = evaluate(e.opt(1))
                var i = 2
                while (i + 1 < e.length() - 1) {
                    val labels = e.opt(i)
                    val matched = if (labels is JSONArray) {
                        (0 until labels.length()).any { eq(input, norm(labels.opt(it))) }
                    } else {
                        eq(input, norm(labels))
                    }
                    if (matched) {
                        return evaluate(e.opt(i + 1))
                    }
                    i += 2
                }
                evaluate(e.opt(e.length() - 1))
            }
            "coalesce" -> (1 until e.length()).asSequence().map { evaluate(e.opt(it)) }.firstOrNull { it != null }
            else -> throw IllegalArgumentException("unsupported expression operator '$op'")
        }
    }

    private fun norm(v: Any?): Any? = if (v == null || v === JSONObject.NULL) null else v

    private fun truthy(v: Any?): Boolean = when (v) {
        null -> false
        is Boolean -> v
        is Number -> v.toDouble() != 0.0
        is String -> v.isNotEmpty()
        else -> true
    }

    private fun eq(a: Any?, b: Any?): Boolean {
        if (a is Number && b is Number) {
            return a.toDouble() == b.toDouble()
        }
        return a == b
    }

    private fun compare(a: Any?, b: Any?): Int {
        if (a is Number && b is Number) {
            return a.toDouble().compareTo(b.toDouble())
        }
        if (a is String && b is String) {
            return a.compareTo(b)
        }
        throw IllegalArgumentException("cannot compare $a and $b")
    }
}

File: ios/RailwayMapCore/Sources/RailwayMapCore/StyleBuilder.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Joel Haasnoot

import Foundation
import os

/**
 The generated map assets shared with the Android app (`android/app/src/main/assets`): styles,
 legend, sprites, glyphs and the world overview. The styles refer to sprites and glyphs with
 Android `asset://` URLs; on iOS those are resolved to files under `root`.
 */
public struct MapAssets: Sendable {
    public let root: URL

    public init(root: URL) {
        self.root = root
    }

    public var ormStyle: URL {
        root.appendingPathComponent("style/orm-style.json")
    }

    public var basemapStyle: URL {
        root.appendingPathComponent("style/basemap-style.json")
    }

    public var legend: URL {
        root.appendingPathComponent("style/legend.json")
    }

    /// The bundled world overview; nil when it is missing from the build.
    public var worldFile: URL? {
        let file = root.appendingPathComponent("world/world.pmtiles")
        return FileManager.default.fileExists(atPath: file.path) ? file : nil
    }

    /// `asset://sprites/symbols` -> `file:///…/assets/sprites/symbols`; other URLs are returned unchanged.
    public func resolve(_ url: String) -> String {
        let scheme = "asset://"
        guard url.hasPrefix(scheme) else {
            return url
        }
        var base = root.absoluteString
        if !base.hasSuffix("/") {
            base += "/"
        }
        return base + url.dropFirst(scheme.count)
    }
}

/**
 Builds the MapLibre style JSON shown on the map.

 The upstream OpenRailwayMap style is a single style that switches between "Infrastructure",
 "Speed", ... via `global-state` expressions. MapLibre Native does not evaluate those, so we
 substitute the chosen state values into the expressions ourselves, evaluate each layer's
 `visibility` expression to a constant and drop hidden layers.

 Every installed country pack gets its own copy of the sources and layers, pointing at the
 PMTiles files on disk. Underneath sits a low-zoom world overview bundled with the app (see
 `WorldLayers`). Thread-safe; building is meant to run off the main thread.
 */
public final class StyleBuilder: @unchecked Sendable {
    private static let log = Logger(subsystem: "com.offlinerailmap.ios", category: "StyleBuilder")

    public let assets: MapAssets
    private let lock = NSLock()
    private var loaded: (orm: JSON, basemap: JSON)?
    private var legendData: JSON?
    private var ormLayersCache: (key: OrmKey, value: OrmLayers)?

    private struct OrmKey: Equatable {
        let mode: MapMode
        let options: MapOptions
        let year: Int
    }

    public init(assets: MapAssets) {
        self.assets = assets
    }

    private func load() throws -> (orm: JSON, basemap: JSON) {
        lock.lock()
        defer { lock.unlock() }
        if let loaded {
            return loaded
        }
        let result = (orm: try JSON(contentsOf: assets.ormStyle), basemap: try JSON(contentsOf: assets.basemapStyle))
        loaded = result
        return result
    }

    /// The legend.json object for a view (`{countries, sourceLayers}`).
    public func legendView(modeId: String) throws -> JSON {
        lock.lock()
        defer { lock.unlock() }
        if legendData == nil {
            legendData = try JSON(contentsOf: assets.legend)
        }
        return legendData?[modeId] ?? [:]
    }

    public func build(mode: MapMode, options: MapOptions, packs: [InstalledPack]) throws -> String {
        let (orm, base) = try load()

        var sources: [String: JSON] = [:]
        var layers: [JSON] = []

        // --- Basemap ---
        // Order: background, world overview, mask over downloaded regions, detailed pack basemaps.
        let baseLayers = base["layers"]?.array ?? []
        let backgroundColor = baseLayers.first { $0["type"]?.string == "background" }?["paint"]?["background-color"]?.string
            .flatMap { $0.isEmpty ? nil : $0 } ?? "#f4f1ec"
        layers.append(contentsOf: baseLayers.filter { !$0.has("source") })

        var worldLayers: [JSON] = []
        if let world = assets.worldFile {
            sources[WorldLayers.source] = WorldLayers.source(pmtilesUrl: Self.pmtilesUrl(world))
            worldLayers = baseLayers.filter { $0.has("source") }.compactMap(WorldLayers.worldCopy)
        }
        let worldLabels = worldLayers.filter { $0["type"]?.string == "symbol" }
        layers.append(contentsOf: worldLayers.filter { $0["type"]?.string != "symbol" })

        let basemapPacks = packs.filter { $0.basemapFile != nil }
        let mask = WorldLayers.maskGeometry(basemapPacks.map(\.info))
        if let mask {
            sources[WorldLayers.maskSource] = WorldLayers.maskSource(mask)
            layers.append(WorldLayers.maskLayer(color: backgroundColor))
        }

        let basemapSource = base["sources"]?["basemap"] ?? [:]
        for pack in basemapPacks {
            var source = basemapSource
            source["url"] = .string(Self.pmtilesUrl(pack.basemapFile!))
            sources["basemap__\(pack.info.id)"] = source
        }
        for layer in baseLayers where layer.has("source") {
            let id = layer["id"]?.string ?? ""
            for pack in basemapPacks {
                var copy = layer
                copy["id"] = .string("\(id)__\(pack.info.id)")
                copy["source"] = .string("basemap__\(pack.info.id)")
                if id == "place-country", let field = copy["layout"]?["text-field"] {
                    // Match the English country names of the world overview next door.
                    copy["layout"]?["text-field"] = WorldLayers.preferEnglish(field)
                }
                layers.append(copy)
            }
        }
        for label in worldLabels {
            layers.append(mask.map { WorldLayers.hideInsideMask(label, geometry: $0) } ?? label)
        }

        // --- OpenRailwayMap ---
        let ormSources = orm["sources"]?.object ?? [:]
        for pack in packs {
            for (name, source) in ormSources where source["type"]?.string == "vector" {
                var copy = source
                copy["url"] = .string(Self.pmtilesUrl(pack.railwayFile))
                sources["\(name)__\(pack.info.id)"] = copy
            }
        }
        let processed = try ormLayers(mode: mode, options: options)
        for layer in processed.layers {
            let id = layer["id"]?.string ?? ""
            let source = layer["source"]?.string ?? ""
            for pack in packs {
                var copy = layer
                copy["id"] = .string("\(id)__\(pack.info.id)")
                copy["source"] = .string("\(source)__\(pack.info.id)")
                layers.append(copy)
            }
        }

        let style: JSON = [
            "version": 8,
            "name": .string("OpenRailwayMap offline · \(mode.label)"),
            "sprite": .string(processed.sprite),
            "glyphs": .string(processed.glyphs),
            "sources": .object(sources),
            "layers": .array(layers),
        ]
        Self.log.info("built style '\(mode.id)' for \(packs.count) pack(s): \(layers.count) layers, \(processed.hidden) upstream layers hidden")
        return style.serialized()
    }

    /// The upstream layers that are visible for a view and options, with `global-state` resolved.
    public struct OrmLayers: Sendable {
        public let layers: [JSON]
        public let hidden: Int
        /// Layers whose visibility did not reduce to a constant and are shown regardless.
        public let unresolved: Int
        public let state: [String: JSON]
        public let sprite: String
        public let glyphs: String
    }

    public func ormLayers(mode: MapMode, options: MapOptions) throws -> OrmLayers {
        let year = Calendar.current.component(.year, from: Date())
        let key = OrmKey(mode: mode, options: options, year: year)
        lock.lock()
        if let cache = ormLayersCache, cache.key == key {
            lock.unlock()
            return cache.value
        }
        lock.unlock()

        let (orm, _) = try load()
        let state = Self.globalState(orm: orm, mode: mode, options: options, year: year)
        let vectorSources = Set((orm["sources"]?.object ?? [:]).filter { $0.value["type"]?.string == "vector" }.keys)
        var result: [JSON] = []
        var hidden = 0
        var unresolved = 0
        for layer in orm["layers"]?.array ?? [] {
            guard let source = layer["source"]?.string, vectorSources.contains(source) else {
                continue
            }
            var substituted = Self.simplify(Self.substitute(layer, state: state))
            let visibility = substituted["layout"]?["visibility"]
            if case .array = visibility {
                Self.log.warning("visibility of \(layer["id"]?.string ?? "?") did not reduce to a constant")
                substituted["layout"]?["visibility"] = "visible"
                unresolved += 1
            }
            if substituted["layout"]?["visibility"]?.string == "none" {
                hidden += 1
                continue
            }
            // Filters that reduced to a constant: drop the layer or drop the filter.
            if case .bool(let keep) = substituted["filter"] {
                if !keep {
                    hidden += 1
                    continue
                }
                substituted.remove("filter")
            }
            result.append(substituted)
        }
        let value = OrmLayers(
            layers: result,
            hidden: hidden,
            unresolved: unresolved,
            state: state,
            sprite: assets.resolve(orm["sprite"]?.string ?? ""),
            glyphs: assets.resolve(orm["glyphs"]?.string ?? "")
        )
        lock.lock()
        ormLayersCache = (key, value)
        lock.unlock()
        return value
    }

    static func pmtilesUrl(_ file: URL) -> String {
        "pmtiles://file://" + file.path
    }

    static func globalState(orm: JSON, mode: MapMode, options: MapOptions, year: Int) -> [String: JSON] {
        var state: [String: JSON] = [:]
        for (key, definition) in orm["state"]?.object ?? [:] {
            state[key] = definition["default"] ?? .null
        }
        state["style"] = .string(mode.id)
        state.merge(mode.globalState) { _, new in new }
        state.merge(options.toGlobalState()) { _, new in new }
        state["theme"] = "light"
        state["pitched"] = false
        state["bearing"] = 0
        state["hillshade"] = false
        state["openHistoricalMap"] = false
        state["date"] = .number(Double(year))
        state["allDates"] = false
        return state
    }

    /// Replaces every `["global-state", key]` with the value of that key.
    static func substitute(_ value: JSON, state: [String: JSON]) -> JSON {
        switch value {
        case .array(let items):
            if items.count == 2, items[0] == "global-state" {
                return items[1].string.flatMap { state[$0] } ?? .null
            }
            return .array(items.map { substitute($0, state: state) })
        case .object(let values):
            return .object(values.mapValues { substitute($0, state: state) })
        default:
            return value
        }
    }

    /**
     Constant-folds sub-expressions whose operands are all constants (after `global-state`
     substitution), so that e.g. `["==", "usage", "usage"]` becomes `true` and `["all", ...]`,
     `["any", ...]` and `["case", ...]` collapse. Besides shrinking the style, this is required
     because MapLibre Native parses a filter such as `["==", "a", "b"]` as a *legacy* filter
     (property "a" equals "b") and rejects `["==", true, false]` outright.
     */
    static func simplify(_ value: JSON) -> JSON {
        if case .object(let values) = value {
            return .object(values.mapValues(simplify))
        }
        guard case .array(let items) = value, let op = items.first?.string else {
            return value
        }
        if op == "literal" {
            return value
        }
        let args = items.dropFirst().map(simplify)
        let rebuilt = JSON.array([.string(op)] + args)
        switch op {
        case "all":
            var remaining: [JSON] = []
            for arg in args {
                if arg == false {
                    return false
                }
                if arg != true {
                    remaining.append(arg)
                }
            }
            switch remaining.count {
            case 0:
                return true
            case 1:
                return remaining[0]
            default:
                return .array(["all"] + remaining)
            }
        case "any":
            var remaining: [JSON] = []
            for arg in args {
                if arg == true {
                    return true
                }
                if arg != false {
                    remaining.append(arg)
                }
            }
            switch remaining.count {
            case 0:
                return false
            case 1:
                return remaining[0]
            default:
                return .array(["any"] + remaining)
            }
        case "case":
            var kept: [JSON] = []
            var i = 0
            while i + 1 < args.count {
                let condition = args[i]
                if condition == true {
                    return kept.isEmpty ? args[i + 1] : .array(["case"] + kept + [args[i + 1]])
                }
                if condition != false && condition != .null {
                    kept.append(condition)
                    kept.append(args[i + 1])
                }
                i += 2
            }
            let fallback = args.last ?? .null
            return kept.isEmpty ? fallback : .array(["case"] + kept + [fallback])
        case "!":
            if case .bool(let b)? = args.first {
                return .bool(!b)
            }
            return rebuilt
        case "==", "!=", "<", "<=", ">", ">=", "in", "match", "coalesce":
            let allConstant = args.enumerated().allSatisfy { index, arg in
                // The label lists of a match (every other argument before the fallback) are data, not expressions.
                if op == "match" && index % 2 == 1 && index < args.count - 1 {
                    return true
                }
                guard case .array(let inner) = arg else {
                    return true
                }
                return inner.first == "literal"
            }
            if allConstant, let result = try? evaluate(rebuilt) {
                return result
            }
            return rebuilt
        default:
            return rebuilt
        }
    }

    // MARK: - A very small expression evaluator, enough for the upstream `visibility` expressions

    struct UnsupportedExpression: Error {}

    static func evaluate(_ e: JSON) throws -> JSON {
        guard case .array(let items) = e, let op = items.first?.string else {
            return e
        }
        func arg(_ index: Int) throws -> JSON {
            try evaluate(index < items.count ? items[index] : .null)
        }
        switch op {
        case "literal":
            return items.count > 1 ? items[1] : .null
        case "case":
            var i = 1
            while i + 1 < items.count {
                if truthy(try arg(i)) {
                    return try arg(i + 1)
                }
                i += 2
            }
            return try arg(items.count - 1)
        case "all":
            for i in 1..<items.count where !truthy(try arg(i)) {
                return false
            }
            return true
        case "any":
            for i in 1..<items.count where truthy(try arg(i)) {
                return true
            }
            return false
        case "!":
            return .bool(!truthy(try arg(1)))
        case "==":
            return .bool(try arg(1) == arg(2))
        case "!=":
            return .bool(try arg(1) != arg(2))
        case "<":
            return .bool(try compare(arg(1), arg(2)) < 0)
        case "<=":
            return .bool(try compare(arg(1), arg(2)) <= 0)
        case ">":
            return .bool(try compare(arg(1), arg(2)) > 0)
        case ">=":
            return .bool(try compare(arg(1), arg(2)) >= 0)
        case "in":
            let needle = try arg(1)
            switch try arg(2) {
            case .array(let haystack):
                return .bool(haystack.contains(needle))
            case .string(let haystack):
                if case .string(let text) = needle {
                    return .bool(haystack.contains(text))
                }
                return false
            default:
                return false
            }
        case "match":
            let input = try arg(1)
            var i = 2
            while i + 1 < items.count - 1 {
                let labels = items[i]
                let matched: Bool
                if case .array(let options) = labels {
                    matched = options.contains(input)
                } else {
                    matched = labels == input
                }
                if matched {
                    return try arg(i + 1)
                }
                i += 2
            }
            return try arg(items.count - 1)
        case "coalesce":
            for i in 1..<items.count {
                let value = try arg(i)
                if value != .null {
                    return value
                }
            }
            return .null
        default:
            throw UnsupportedExpression()
        }
    }

    private static func truthy(_ value: JSON) -> Bool {
        switch value {
        case .null:
            return false
        case .bool(let b):
            return b
        case .number(let n):
            return n != 0
        case .string(let s):
            return !s.isEmpty
        default:
            return true
        }
    }

    private static func compare(_ a: JSON, _ b: JSON) throws -> Int {
        switch (a, b) {
        case (.number(let x), .number(let y)):
            return x < y ? -1 : (x > y ? 1 : 0)
        case (.string(let x), .string(let y)):
            return x < y ? -1 : (x > y ? 1 : 0)
        default:
            throw UnsupportedExpression()
        }
    }
}
